addons/mail/tests/test_mail_gateway.py

```diff
@@ -366,6 +510,51 @@ def test_message_parse(self):
         self.assertEqual(res['body'], '')
         self.assertEqual(res['attachments'][0][0], 'thetruth.pdf')
 
+    @mute_logger('odoo.addons.mail.models.mail_thread')
+    def test_message_parse_eml(self):
+        """ Test parsing of mail with embedded emails as eml(msg) files.
+        """
+
+        # Tree structure of the message:
+        def display(message_as_tree):
+            return d(message_as_tree, "", "")
+
+        def d(tree, prefix, pprint_tree):
+            bound = tree.get_boundary().replace("-", "") if tree.get_boundary() else ""
+            content_type = tree.get_content_type() or ""
+            filename = tree.get_filename() or ""
+            content = "STR" if tree.get_payload(decode=True) else "NON"
+            m = ", ".join([bound, content_type, filename, content])
+            pprint_tree += prefix + m + "\n"
+            if isinstance(tree.get_payload(), str):  # a fucking leaf
+                pass
+            else:
+                for son in tree.get_payload():
+                    pprint_tree = d(son, "----" + prefix, pprint_tree)
+            return pprint_tree
+
+        """ display(res) (i.e. MAIL_EML_ATTACHMENT) results in:
```

Comment:
It seems to only assert is totally not linked to the display method. What is its purpose ?

Reply:
The purpose of the display is to see the tree structure of the mail.
As explained, the filename is currently not extracted because the parsing does not take into account the structure. 
To get rid of the TODO, you would need to actually take it into account, so this display is to make that explicit.

Reply:
My question seems still unanswered. Unless I am mistaken it is not called and therefore unnecessary.

Reply:
Currently the parsing does not get the filename of the attachment because it doesn't take into account the tree structure of the email.
To be able to do it, you would have to use this tree structure. 
This function helps visualize the tree structure to see how things should be done.

Since this bug would require a massive rewrite of the parsing, it cannot be done in stable. 
However if somebody wants to rewrite this correctly, it is a little hint to get it right.

Reply:
Ok. Then I would better have a commit message explaining how it should be done completely, why it cannot be done completely, and remove this code, especially it is used for debug purpose and is not part of the parsing. I think commit message is the right place to explain things. If you think we could improve it in master, feel free to create a task with complete explanations so that we can work on it. Thanks !

Reply:
Fine, I now see the errors of my way. I won't bother you with this anymore.